src/compiler/checker.ts

```diff
@@ -3945,6 +3981,12 @@ namespace ts {
                             }
                         }
                         result.push(getSignatureFromDeclaration(<SignatureDeclaration>node));
+                        break;
+
+                    case SyntaxKind.PropertyAccessExpression:
+                        // Inferred class method
+                        result = getSignaturesOfType(checkExpressionCached((<BinaryExpression>node.parent).right), SignatureKind.Call);

```

Comment:
I'm concerned with using `checkExpression` here. It might recursively get you back to this same spot. We're just supposed to create a signature from a declaration here.